src/lex.js

```diff
@@ -567,92 +590,108 @@ Lexer.prototype = {
     var index = 0;
     var type, char;
 
-    function isNonAsciiIdentifierStart(code) {
-      return nonAsciiIdentifierStartTable.indexOf(code) > -1;
+    function isValidIdentifierCharacter(code, position) {
+      if (position === "start") {
+        return asciiData.start[code] || nonAsciiIdentifierStart.BMP.indexOf(code) > -1 ||
+          (code > 0xffff && nonAsciiIdentifierStart.astral.indexOf(code) > -1);
+      } else {
+        return isValidIdentifierCharacter(code, "start") || asciiData.part[code] ||
+          nonAsciiIdentifierPartOnly.BMP.indexOf(code) > -1 ||
+          (code > 0xffff && nonAsciiIdentifierPartOnly.astral.indexOf(code) > -1);
+      }
     }
 
-    function isNonAsciiIdentifierPart(code) {
-      return isNonAsciiIdentifierStart(code) || nonAsciiIdentifierPartTable.indexOf(code) > -1;
+    function isHexNumber(str) {
+      return (/^[0-9a-fA-F]+$/).test(str);
     }
 
-    function isHexDigit(str) {
-      return (/^[0-9a-fA-F]$/).test(str);
-    }
-
-    var readUnicodeEscapeSequence = function() {
+    var readUnicodeEscapeSequence = function(position) {
       /*jshint validthis:true */
-      index += 1;
-
-      if (this.peek(index) !== "u") {
+      if (this.peek(index + 1) !== "u") {
         return null;
       }
 
-      var ch1 = this.peek(index + 1);
-      var ch2 = this.peek(index + 2);
-      var ch3 = this.peek(index + 3);
-      var ch4 = this.peek(index + 4);
-      var code;
-
-      if (isHexDigit(ch1) && isHexDigit(ch2) && isHexDigit(ch3) && isHexDigit(ch4)) {
-        code = parseInt(ch1 + ch2 + ch3 + ch4, 16);
+      var hexCode = "";
+      var escapeSequence = "\\u";
 
-        if (asciiIdentifierPartTable[code] || isNonAsciiIdentifierPart(code)) {
-          index += 5;
-          return "\\u" + ch1 + ch2 + ch3 + ch4;
+      if (this.peek(index + 2) === "{") { // \u{...}
+        if (!state.inESNext(true)) {
+          this.trigger("warning", {
+            code: "W119",
+            line: this.line,
+            character: this.char + index,
+            data: [ "\\u{...}" ]
```

Comment:
What message should JSHint warn?
Is `'\u{...}' is only available in ES6 (use esnext option).` ok?